### PR TITLE
tables: Fix macOS signature table reporting

### PR DESCRIPTION
**Pull request by theopolis**
This addresses CVE-2018-6336.

I am just the messenger opening the PR here, @obelisk is the author of the fix and the one who found the way to implement the correct checking.

### CODE REVIEW

File: osquery/tables/system/darwin/signature.mm [hunk 1]
```diff
@@ -58,130 +62,159 @@ Status getVerifyFlags(SecCSFlags& flags) {
...
-      r["team_identifier"] = "";
-      CFTypeRef teamIdent = nullptr;
-      if (CFDictionaryGetValueIfPresent(codeInfo, kSecCodeInfoTeamIdentifier,
-                                        &teamIdent)) {
-        r["team_identifier"] = stringFromCFString((CFStringRef)teamIdent);
+  if (ident == nullptr) {
+    CFRelease(code_info);
+    results.push_back(r);
+    CFRelease(static_code);
+    return Status(1, "No identifier found for file(" + arch + "): " + path);
```
Comment (theopolis) [inline osquery/tables/system/darwin/signature.mm]: I feel like this error is not consistent with the above. I'd recommend just reporting that the identifier is not found for the arch.

File: osquery/tables/system/darwin/signature.mm [hunk 2]
```diff
@@ -58,130 +62,159 @@ Status getVerifyFlags(SecCSFlags& flags) {
...
+  if (hashInfo != nullptr) {
+    r["cdhash"].reserve(CC_SHA1_DIGEST_LENGTH);
+    // Get the SHA-1 bytes
+    std::stringstream ss;
+    auto bytes = CFDataGetBytePtr(hashInfo);
+    if (bytes != nullptr &&
+        CFDataGetLength(hashInfo) == CC_SHA1_DIGEST_LENGTH) {
+      // Write bytes as hex strings
+      for (size_t n = 0; n < CC_SHA1_DIGEST_LENGTH; n++) {
+        ss << std::hex << std::setfill('0') << std::setw(2);
```
Comment (theopolis) [inline osquery/tables/system/darwin/signature.mm]: Maybe there's another function for this?

File: osquery/tables/system/darwin/signature.mm [hunk 3]
```diff
@@ -58,130 +62,159 @@ Status getVerifyFlags(SecCSFlags& flags) {
...
+        std::string commonName;
+        genCommonName(x509_cert, subject, commonName, issuer);
+        r["authority"] = commonName;
+        X509_free(x509_cert);
+      } else {
+        VLOG(1) << "Error decoding DER encoded certificate";
       }
-    } else {
-      VLOG(1) << "No identifier found for file: " << path;
+      CFRelease(der_encoded_data);
```
Comment (theopolis) [inline osquery/tables/system/darwin/signature.mm]: This releases if `def_encoded_data == nullptr`

File: osquery/tables/system/darwin/signature.mm [hunk 4]
```diff
@@ -58,130 +62,159 @@ Status getVerifyFlags(SecCSFlags& flags) {
...
-            X509_free(x509_cert);
-          } else {
-            VLOG(1) << "Error decoding DER encoded certificate";
-          }
-          CFRelease(der_encoded_data);
-        }
+  // Team Identifier
+  r["team_identifier"] = "";
+  CFTypeRef teamIdent = nullptr;
+  if (CFDictionaryGetValueIfPresent(
```
Comment (theopolis) [inline osquery/tables/system/darwin/signature.mm]: I feel like this is the best way to use these APIs.
- Check if the value is present and get the `CFTypeRef`.
- Check if the type is the type you expect (using CF APIs for comparison)
- Explicit cast and use that value inside the scope.

File: osquery/tables/system/darwin/signature.mm [hunk 5]
```diff
@@ -58,130 +62,159 @@ Status getVerifyFlags(SecCSFlags& flags) {
...
+      (CFStringRef)CFDictionaryGetValue(code_info, kSecCodeInfoIdentifier);
 
-      // Team Identifier
-      r["team_identifier"] = "";
-      CFTypeRef teamIdent = nullptr;
-      if (CFDictionaryGetValueIfPresent(codeInfo, kSecCodeInfoTeamIdentifier,
-                                        &teamIdent)) {
-        r["team_identifier"] = stringFromCFString((CFStringRef)teamIdent);
+  if (ident == nullptr) {
+    CFRelease(code_info);
```
Comment (theopolis) [inline osquery/tables/system/darwin/signature.mm]: nit, The release order is bothering me, haha!

File: osquery/tables/system/darwin/signature.mm [hunk 6]
```diff
@@ -58,130 +62,159 @@ Status getVerifyFlags(SecCSFlags& flags) {
...
-      // and, since it didn't error above, it's *also* a valid signature.
-      if (isValidated) {
-        r["signed"] = INTEGER(1);
-      }
-      r["identifier"] = stringFromCFString(ident);
+      static_code,
+      kSecCSSigningInformation | kSecCSRequirementInformation,
+      &code_info);
+
+  if (result != errSecSuccess) {
```
Comment (fmanco) [inline osquery/tables/system/darwin/signature.mm]: Is `code_info` guaranteed to be null when the function fails? The documentation is not explicit about it.

File: osquery/tables/system/darwin/signature.mm [hunk 7]
```diff
@@ -58,130 +62,159 @@ Status getVerifyFlags(SecCSFlags& flags) {
...
+  }
+
+  r["identifier"] = stringFromCFString(ident);
+
+  // Get CDHash
+  r["cdhash"] = "";
+  CFDataRef hashInfo =
+      (CFDataRef)CFDictionaryGetValue(code_info, kSecCodeInfoUnique);
+  if (hashInfo != nullptr) {
+    r["cdhash"].reserve(CC_SHA1_DIGEST_LENGTH);
```
Comment (fmanco) [inline osquery/tables/system/darwin/signature.mm]: I know this is not modified code, but why are we doing this reserve? Isn't this string going to be replaced in any case?

File: osquery/tables/system/darwin/signature.mm (hunk 6 above)
Comment (obelisk) [reply osquery/tables/system/darwin/signature.mm]: We can add an extra check here.

File: osquery/tables/system/darwin/signature.mm (hunk 3 above)
Comment (obelisk) [reply osquery/tables/system/darwin/signature.mm]: It does? It's in a block guarded with `if (der_encoded_data != nullptr) {`

**Pull request merged**